src/stats/cdf.rs

```diff
@@ -0,0 +1,209 @@
+        };
+
+        if relative_eq!(cdf.total_prob(), 0.0) && cdf.total_prob() > 0.0 {
+            cdf.inner.last_mut().unwrap().1 = 0.0;
+        }
+
+        cdf
+    }
+
+    /// Create CDF from iterator. This can be used to replace the values of a CDF.
+    pub fn from_cdf<I: Iterator<Item = (T, LogProb)>>(entries: I) -> Self {
+        CDF { inner: entries.collect_vec() }
+    }
+
+    /// Reduce CDF by omitting values with zero probability.
+    pub fn reduce(self) -> Self {
+        let mut inner = Vec::new();
+        let mut last = f64::NEG_INFINITY;
+        for e in self.inner.into_iter() {
+            if last != e.1 {
+                last = e.1;
+                inner.push(e);
+            }
+        }
+        CDF { inner: inner }
+    }
+
+    /// Downsample CDF to n entries.
+    pub fn sample(mut self, n: usize) -> Self {
+        if self.inner.len() <= n {
+            self
+        }
+        else {
+            let s = self.inner.len() / (n - 1);
+            let last = self.inner.pop().unwrap();
+            let mut inner = self.inner.into_iter().step(s).collect_vec();
+            inner.push(last);
+            CDF {
+                inner: inner
+            }
+        }
+    }
+
+    /// Provide iterator.
+    pub fn iter(&self) -> slice::Iter<(T, f64)>{
+        self.inner.iter()
+    }
+
+    /// Iterator over corresponding PMF.
+    pub fn iter_pmf<'a>(&'a self) -> CDFPMFIter<'a, T> {
+        fn cdf_to_pmf<'a, G>(last_prob: &mut LogProb, e: &'a (G, LogProb)) -> Option<(&'a G, LogProb)> {
+            let &(ref value, cdf_prob) = e;
+            let prob = logprobs::sub(cdf_prob, *last_prob);
+            *last_prob = cdf_prob;
+            Some((value, prob))
+        }
+        self.inner.iter().scan(f64::NEG_INFINITY, cdf_to_pmf)
+    }
+
+    /// Get cumulative probability for a given value. If the value is not present,
+    /// return the probability of the previous value. Complexity O(log n).
+    pub fn get(&self, value: &T) -> Option<LogProb> {
+        if self.inner.is_empty() {
+            None
+        }
+        else {
+            Some(match self.inner.binary_search_by(|e| e.0.partial_cmp(value).unwrap()) {
+                Ok(i) => self.inner[i].1,
+                Err(i) => if i > 0 { self.inner[i - 1].1 } else { f64::NEG_INFINITY }
+            })
+        }
+    }
+
+    /// Get probability (i.e. probability mass) for a given value. Complexity O(log n).
+    pub fn get_pmf(&self, value: &T) -> Option<LogProb> {
+        if self.inner.is_empty() {
+            None
+        }
+        else {
+            Some(match self.inner.binary_search_by(|e| e.0.partial_cmp(value).unwrap()) {
+                Ok(i) => if i > 0 { logprobs::sub(self.inner[i].1, self.inner[i - 1].1) } else { self.inner[0].1 },
+                Err(i) => if i > 0 { self.inner[i - 1].1 } else { f64::NEG_INFINITY }
+            })
+        }
+    }
+
+    /// Return total probability.
+    pub fn total_prob(&self) -> LogProb {
```

Comment:
You have checks above for `self.inner` being empty, but here it seems like you expect it to be non-empty. Emptiness should either be guarded against in the creation and mutation of `self.inner`, or you should handle it gracefully here (at least an `expect` rather than `unwrap`).